builtin/v12/migration/miner.go

```diff
@@ -224,65 +271,112 @@ delta = prevInRoot - inRoot
 	diffs, err := amt.Diff(ctx, store, store, prevInRoot, inRoot, amt.UseTreeBitWidth(miner11.SectorsAmtBitwidth))
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to diff old and new Sector AMTs: %w", err)
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to diff old and new Sector AMTs: %w", err)
 	}
 
 	inSectors, err := miner11.LoadSectors(store, inRoot)
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
 	}
-
 	prevOutSectors, err := miner12.LoadSectors(store, prevOutRoot)
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+	}
+
+	// load previous HAMT sector index for this specific minerAddr
+	sectorToDealIdHamt, err := builtin.LoadTree(store, prevSectorIndexRoot)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("creating new state tree: %w", err)
 	}
 
 	for _, change := range diffs {
 		switch change.Type {
 		case amt.Remove:
 			if err := prevOutSectors.Delete(change.Key); err != nil {
-				return cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+			}
+
+			//remove sector from HAMT index
+			err = removeSectorNumberToDealIdFromHAMT(sectorToDealIdHamt, change.Key, store)
+			if err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to remove sector from HAMT: %w", err)
 			}
+
 		case amt.Add:
 			fallthrough
 		case amt.Modify:
 			sectorNo := abi.SectorNumber(change.Key)
 			info, found, err := inSectors.Get(sectorNo)
 			if err != nil {
-				return cid.Undef, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
 			}
 
 			if !found {
-				return cid.Undef, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
+				return cid.Undef, cid.Undef, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
 			}
 
-			if err := prevOutSectors.Set(change.Key, migrateSectorInfo(*info)); err != nil {
-				return cid.Undef, xerrors.Errorf("failed to set migrated sector %d in prevOutSectors", sectorNo)
+			if err := prevOutSectors.Set(change.Key, migrateSectorInfo(*info, store)); err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to set migrated sector %d in prevOutSectors", sectorNo)
+			}
+
+			// add sector to the HAMT
+			err = addSectorNumberToDealIdHAMT(sectorToDealIdHamt, *info, store)
+			if err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to add sector %d to HAMT: %w", sectorNo, err)
 			}
 		}
 	}
 
-	return prevOutSectors.Root()
+	//return the hAmt
```

Comment:
nit: Either remove or be more specific that the hamt is the sector id => deal id mapping

Reply:
//return the sector to deal id Hamt